### DISCUSSION TITLE
State change from "then" method does not update UI in StateNotifier

### DISCUSSION THREAD

**Discussion by cobalt-60** [rrousselGit/riverpod#1798] [Q&A]
I display a UI and load some data asynchronously, currently from SharedPreferences, and want the data to be displayed, once it's loaded.

This is how my StateNotifier looks like:

```
abstract class BaseService<S> extends StateNotifier<S> {
  final StateNotifierProviderRef<BaseService<S>, S> ref;
  final Future<SharedPreferences> _fPrefs;
  SharedPreferences? _prefs;
  String sharedPreferencesKey;

  BaseService(super.state, this.sharedPreferencesKey, this.ref) : _fPrefs = SharedPreferences.getInstance() {
    _fPrefs.then((value) {
      _prefs = value;
      String? json = _prefs?.getString(sharedPreferencesKey);
      print("BaseService:loading:\n ${((json == null) ? "null" : json)}");
      if (json != null) {
        ref.notifier.state = fromJson(json);
      }
    });
  }

  @override
  bool updateShouldNotify(S old, S current) {
    print("BaseService:updateShouldNotify: old=$old current=$current");
    // return super.updateShouldNotify(old, current);
    return true;
  }
}
```

And the example for the User object

```
class UserService extends BaseService<User> {

  UserService(StateNotifierProviderRef<UserService, User> ref) : super(User(), APP_STATE_USER_KEY, ref);
}

final userStateProvider = StateNotifierProvider<UserService, User>((ref) => UserService(ref));

```

In the log I can see how the UI is set up with an empty User, which is fine, and then the BaseService loading the User from SharedPreferences. 

But whatever I try, the UI is not updated with the user data.
I added the ref parameter exactly for that, but it didn't change anything.
And I see the "updateShouldNotify" method get called with the correct old and new state, but the UI doesn't change.

Anyone any ideas?

**Comment by rrousselGit** [comment]
If `updateShouldNotify` is called, I'd question how your UI is listening to the state instead.
That could be the issue

**Reply by cobalt-60** [reply]
All other state updates are processed correctly. Whenever the state is changed from UI callbacks like onPressed it works fine.

**Reply by cobalt-60** [reply]
I suppose, I'm accessing the wrong instance of the StateNotifier. I thought the ref parameter is providing you exactly this, the instance that Riverpod is using internally. I will do some additional debugging here

**Comment by cobalt-60** [comment]
What I see now, is that I don't access the notifier with the read() method via the StateNotifierProvider.

I tried a bloody hack, to return the specific instance of the StateNotifierProvider via an overloaded method (getStateNotifierProvider()) from the Service implementations like the UserService. And then set the state like this in the BaseState ctor:

`ref.read(getStateNotifierProvider().notifier).state = fromJson(json);
`

But Riverpod friendly reminds me, not to to try this again ("A provider cannot depend on itself").

So I'm running out of ideas again.
